include/oneapi/dpl/internal/distributed_ranges_impl/sp/algorithms/fill.hpp

```diff
@@ -0,0 +1,115 @@
+// -*- C++ -*-
+//===----------------------------------------------------------------------===//
+//
+// Copyright (C) Intel Corporation
+//
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+// This file incorporates work covered by the following copyright and permission
+// notice:
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef _ONEDPL_DR_SP_FILL_HPP
+#define _ONEDPL_DR_SP_FILL_HPP
+
+#include <memory>
+#include <sycl/sycl.hpp>
+#include <type_traits>
+
+#include "../../detail/segments_tools.hpp"
+#include "../detail.hpp"
+#include "../device_ptr.hpp"
+#include "../util.hpp"
+
+namespace oneapi::dpl::experimental::dr::sp
+{
+
+template <std::contiguous_iterator Iter>
+requires(!std::is_const_v<std::iter_value_t<Iter>> && std::is_trivially_copyable_v<std::iter_value_t<Iter>>) sycl::event
+    fill_async(Iter first, Iter last, const std::iter_value_t<Iter>& value)
```

Comment:
Firstly, I would ask why the implementation of "fill" algo is not consistent with others, `for_each ` for example?
1) Why there  is no  `oneapi::dpl::experimental::dr::sp::for_each_async`, but there is `fill_async`?
2) Why sync version `fill` doesn't have ExecutionPolicy parameter, but `for_each` has one?
What exactly versions (sync/async) and signatures do we want to release?

Reply:
We originally implemented fill using SYCL `fill`.  The asynchronous version of fill is a single-GPU algorithm that originally called SYCL's `q.fill()`.  However, SYCL's fill fails on buffers larger than 2^31, so we implemented a work-around.

In the future we'd perhaps like to go back to SYCL's `q.fill()`, as in theory you might be able to implement a `fill` faster than with a `for_each`.  However, right now, it's pretty similar to `for_each`, and we could probably just implement it directly with `for_each`.

Lack of execution policy is just an oversight.